types/block.go

```diff
@@ -413,6 +413,9 @@ type Data struct {
 
 // Hash returns the hash of the data
 func (data *Data) Hash() cmn.HexBytes {
+	if data == nil {
```

Comment:
I'd argue that such things lead to "hard to debug" errors. One day we'll have a bug with the title "Don't understand why tx hash is empty." Instead of adding `X == nil` everywhere, I think we should locate/change the logic in places where `nil` might come from (basically scan every usage).

Reply:
Being this close to launch, or even for that matter at anytime, I'd rather avoid nil panics, moreover for untested paths than an endless debug. There have been many runtime panics reported by users and at times @ebuchman has wished we'd write everything in Rust to avoid such.

Reply:
Or at least enhance the comment: "Hash returns the hash of the data. If given data was nil it foobarblub"

Reply:
With @melekes here, if `Hash()` can't satisfy it's contract it should make this explicit and return an additional `error` which can be used by the caller for control flow.

Reply:
Let's delay launch until we re-write the whole thing in rust :stuck_out_tongue_winking_eye:

The `block.Hash` already has these nil checks. I agree it'd be nice not to have to do them, but if we're going to do some, we should probably do them everywhere (in this case, `Commit.Hash` included!).

That said, we should certainly do a better job of clearly articulating/noting under what circumstances these values can be nil.